### PR TITLE
[WIP] add builder spec validator registration structs to the v1 package

### PR DESCRIPTION
**Pull request by ciaranmcveigh5**
**Resolves Issues**

#18 

**Description**

Adding builder api validator registration to the go-eth2-client

### CODE REVIEW

File: api/v1/validatorregistration.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+	"github.com/attestantio/go-eth2-client/spec/phase0"
+	"github.com/goccy/go-yaml"
+	"github.com/pkg/errors"
+)
+
+// Validator Registration represents a ValidatorRegistrationV1.
+type ValidatorRegistration struct {
+	FeeRecipient bellatrix.ExecutionAddress
+	GasLimit     uint64
+	Timestamp    uint64
```
Comment (mcdee) [inline api/v1/validatorregistration.go:34]: This should be a `time.Time` (as used in `genesis.go`).

File: api/versionedsignedvalidatorregistration.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	apiv1 "github.com/attestantio/go-eth2-client/api/v1"
+	"github.com/attestantio/go-eth2-client/spec"
+	"github.com/attestantio/go-eth2-client/spec/bellatrix"
+	"github.com/attestantio/go-eth2-client/spec/phase0"
+)
+
+// VersionedSignedValidatorRegistration contains a versioned SignedValidatorRegistrationV1.
+type VersionedSignedValidatorRegistration struct {
+	Version   spec.DataVersion
+	Bellatrix *apiv1.SignedValidatorRegistration
```
Comment (mcdee) [inline api/versionedsignedvalidatorregistration.go:28]: This object isn't under the consensus specs versioning (this is stated explicitly in https://github.com/ethereum/builder-specs/blob/main/specs/builder.md#independently-versioned) so should use its own versioning (perhaps `spec.BuilderVersion`)

**Review by mcdee — CHANGES_REQUESTED** [review]
Thank you for your submission.  A few items that need tweaking, but nothing major.

File: api/v1/validatorregistration_test.go [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+			err:   "timestamp missing",
+		},
+		{
+			name:  "TimestampWrongType",
+			input: []byte(`{"fee_recipient":"0x000102030405060708090a0b0c0d0e0f10111213","gas_limit":"100","timestamp":true,"pubkey":"0x000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f202122232425262728292a2b2c2d2e2f"}`),
+			err:   "invalid JSON: json: cannot unmarshal bool into Go struct field validatorRegistrationJSON.timestamp of type string",
+		},
+		{
+			name:  "TimestampInvalid",
+			input: []byte(`{"fee_recipient":"0x000102030405060708090a0b0c0d0e0f10111213","gas_limit":"100","timestamp":"invalid","pubkey":"0x000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f202122232425262728292a2b2c2d2e2f"}`),
```
Comment (ciaranmcveigh5) [inline api/v1/validatorregistration_test.go:87]: time.unix() is of type int64 not sure how you want to test for invalid data - for now i've just used "invalid"

**Comment by ciaranmcveigh5** [comment]
Thanks for reviewing have made the requested updates, tests are passing - let me know if the builderVersion implementation isn't in line with what you were thinking

File: api/versionedvalidatorregistration.go [hunk 4]
```diff
@@ -24,7 +25,7 @@ import (
 
 // VersionedValidatorRegistration contains a versioned ValidatorRegistrationV1.
 type VersionedValidatorRegistration struct {
-	Version   spec.DataVersion
+	Version   spec.BuilderVersion
 	Bellatrix *apiv1.ValidatorRegistration
```
Comment (mcdee) [inline api/versionedvalidatorregistration.go:29]: This field should map to the builder version, so be called `V1` rather than `Bellatrix`.

**Review by mcdee — CHANGES_REQUESTED** [review]

**Review by mcdee — APPROVED** [review]

File: spec/builderversion.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+	BuilderVersionV1 BuilderVersion = iota
+)
+
+var responseBuilderVersionStrings = [...]string{
+	"V1",
+}
+
+// MarshalJSON implements json.Marshaler.
+func (d *BuilderVersion) MarshalJSON() ([]byte, error) {
+	return []byte(fmt.Sprintf("%q", responseVersionStrings[*d])), nil
```
Comment (mcdee) [inline spec/builderversion.go:35]: Should be `responseBuilderVersionStrings`

File: spec/builderversion.go [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+		*d = BuilderVersionV1
+	default:
+		err = fmt.Errorf("unrecognised response version %s", string(input))
+	}
+	return err
+}
+
+// String returns a string representation of the
+func (d BuilderVersion) String() string {
+	if int(d) >= len(responseVersionStrings) {
```
Comment (mcdee) [inline spec/builderversion.go:52]: Should be `responseBuilderVersionStrings`

File: spec/builderversion.go [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+		return "unknown"
+	}
+	return responseVersionStrings[d]
```
Comment (mcdee) [inline spec/builderversion.go:55]: Should be `responseBuilderVersionStrings`

**Review by mcdee — CHANGES_REQUESTED** [review]

**Comment by moshe-blox** [comment]
I have made some changes and successfully tested against Teku on Sepolia, please take a look at https://github.com/ciaranmcveigh5/go-eth2-client/pull/1 if you can.

Hopefully this can get merged to this repo eventually.

**Comment by mcdee** [comment]
Need to address the above items where `responseVersionStrings` should be `responseBuilderVersionStrings`, apart from that looking good.

**Comment by ciaranmcveigh5** [comment]
hey @mcdee I believe the above items where responseVersionStrings should be responseBuilderVersionStrings have been addressed in @moshe-blox commits

currently builderversion.go has `responseBuilderVersionStrings`

and dataversion.go has `responseVersionStrings`

**Pull request merged**